fog/overseer/server/src/service.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+use crate::{error::OverseerError, worker::OverseerWorker};
+use mc_common::logger::{log, Logger};
+use mc_fog_recovery_db_iface::RecoveryDb;
+use mc_fog_uri::FogIngestUri;
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc,
+};
+
+// Implements core logic for the Fog Overseer HTTP server.
+pub struct OverseerService<DB: RecoveryDb + Clone + Send + Sync + 'static>
+where
+    OverseerError: From<DB::Error>,
+{
+    ingest_cluster_uris: Vec<FogIngestUri>,
+    logger: Logger,
+    overseer_worker: Option<OverseerWorker>,
+    recovery_db: DB,
+    stop_requested: Arc<AtomicBool>,
+}
+
+impl<DB: RecoveryDb + Clone + Send + Sync + 'static> OverseerService<DB>
+where
+    OverseerError: From<DB::Error>,
+{
+    pub fn new(ingest_cluster_uris: Vec<FogIngestUri>, recovery_db: DB, logger: Logger) -> Self {
+        Self {
+            ingest_cluster_uris,
+            logger,
+            overseer_worker: None,
+            recovery_db,
+            stop_requested: Arc::new(AtomicBool::new(false)),
+        }
+    }
+
+    /// Starts the Fog Overseer worker thread.
+    pub fn start(&mut self) -> Result<(), OverseerError> {
+        let ret = self.start_helper();
+        if let Err(ref err) = ret {
+            log::error!(self.logger, "Stopping ingest server due to {:?}", err);
+            self.stop()?;
+        }
+        ret
+    }
+
+    fn start_helper(&mut self) -> Result<(), OverseerError> {
```

Comment:
Why split this out from start?

Reply:
It helps keep the start function small and focused- it deals with handling any errors that might arise from starting the service without getting into the nitty gritty details of what threads etc. are started. Moreover, splitting the logic out between a `start` and `start_helper` is the convention for Fog services.